src/oss/langchain-middleware.mdx

```diff
@@ -0,0 +1,160 @@
+---
+title: Middleware
+---
+
+## Overview
+
+Middleware is new configuration option available for agents.
+It provides a way to more tightly control what happens inside the agent.
+
+The core agent loop involves calling a `model`, letting it choose `tools` to execute, and then finishing when it calls no more tools..
+Middleware provides way to control what happens before and after those steps.
+
+## Using in an agent
+
+:::python
+You can use middleware in an agent by passing it `create_react agent`:
+```python
+middleware: list[AgentMiddleware] = ...
+agent = create_create_agent(
+    ...,
+    middleware=middleware,
+    ...
+)
+```
+:::
+:::js
+TODO
+:::
+
+Middleware is highly flexible and replaces some other functionality in the agent.
+As such, there are some limitations on the the arguments used to create the agent:
+- `model` must be either a string or a BaseChatModel. Will error if a function is passed. If you want to dynamically control the model, use `AgentMiddleware.modify_model_request`
+- `prompt` must be either a string or None. Will error if a function is passed. If you want to dynamically control the prompt, use `AgentMiddleware.modify_model_request`
+- `pre_model_hook` must not be provided. Will error if provided. Use `AgentMiddleware.before_model` instead
+- `post_model_hook` must not be provided. Will error if provided. Use `AgentMiddleware.after_model` instead
+- `pre_model_hook` must not be provided. Will error if provided. Use `AgentMiddleware.before_model` instead
+
+
+
+## AgentMiddleware
+
+AgentMiddleware currently provides three different ways to modify the core agent loop:
+- `before_model`: runs before the model is run. Can update state or exit early with AgentGoto
+- `modify_model_request`: runs before the model is run. Cannot update state or exit early with AgentGoto
+- `after_model`: runs after the model is run. Can update state or exit early with AgentGoto
+
+### `before_model`
+
+Runs before the model is run. Can modify state by returning a new state object or state update.
+Can return AgentGoto to exit early
+
+Signature:
+:::python
+```python
+def before_model(self, state: AgentState) -> AgentUpdate | AgentGoTo | None:
+```
+:::
+:::js
+TODO
+:::
```

Comment:
```suggestion
:::js
import type { AgentState, AgentUpdate, AgentGoTo } from "langchain";

function beforeModel (state: AgentState): AgentUpdate | AgentGoTo | undefined {
    // ...
}
:::
```

Reply:
the implementation will also support promise return values but including it in this example would make it unnecessarily complex.